README.md
# ledger-thorchain-js

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

This package provides a basic client library to communicate with a THORChain App running in a Ledger Nano S/X

We recommend using the [npm package](https://www.npmjs.com/package/@thorchain/ledger-thorchain) in order to receive updates/fixes.

## Build

```bash
yarn build
```

## Tests

### Unit tests

```bash
yarn test:unit
```

### e2e tests

`e2e' tests are to run some tests against your connected Ledger locally (no CI).

Create a `.e2e.env` (use `.e2e.env.sample` as template) to add `THOR_PUB_KEY` (public key of your Ledger) and `THOR_ADDRESS` (`thor` address of your Ledger).

Connect your Ledger, open `THORChain` app on it and start e2e tests locally:

```bash
yarn test:e2e
```

Hint: Add [`.only`](https://jestjs.io/docs/api#testonlyname-fn-timeout) to the tests to run a single test only


## Example

This repo also includes a simple Vue example for U2F and WebUSB.

To run it locally

```bash
yarn serve
```
and open Chrome at https://localhost:8080/

![Example](docs/example.png)


CHANGELOG.md
# 0.2.0 (2023-12-18)

## Update

- Rewrite for vue 3
- Update deps
- Tested in serve env


# 0.1.0 (2021-XX-XX)

## Update

- Rewrite in TypeScript
- Introduce custom `LedgerError`
- Improve `Promise` handling

## Breaking changes

- Values of misc. `XYZResponse`'s have been renamed to use camel case (instead of snake case)
- Re-organize helper function to be available in `helpers` and `common` only (Deprecated `v1` helpers have been removed)
- Rename some methods of THORChain app (mostly to prefix getter methods with `get...`)

# 0.0.3 (2021-07-20)

## Internal updates

- Update dependencies, Remove deprecated `@ledgerhq/hw-transport-u2f` from examples, Use same `jest.setTimeout` for all `e2e` tests [#2](https://github.com/thorchain/ledger-thorchain-js/pull/2)
- Update tests [#1](https://github.com/thorchain/ledger-thorchain-js/pull/1)

# 0.0.2 (2021-06-30)

Initial release


tests/basic.spec.js
import THORChainApp from "index.js";

test("check address conversion", async () => {
  const pkStr = "0264bff60f78aae3326b4fc3e16a7c48d0158bfe175fa4a9361190a42565662ae6";
  const pk = Buffer.from(pkStr, "hex");
  const addr = THORChainApp.getBech32FromPK("thor", pk);
  expect(addr).toEqual("thor1mwyrp6lj85swy5e5g4hjlaacm33g6rw3p4qmq4");
});


tests/basic.ispec.js
import THORChainApp from "index.js";
import TransportNodeHid from "@ledgerhq/hw-transport-node-hid";
import { expect, test } from "jest";
import secp256k1 from "secp256k1/elliptic";
import crypto from "crypto";
import { ERROR_CODE } from "../src/common";

test("get version", async () => {
  const transport = await TransportNodeHid.create(1000);

  const app = new THORChainApp(transport);
  const resp = await app.getVersion();
  console.log(resp);

  expect(resp.return_code).toEqual(ERROR_CODE.NoError);
  expect(resp.error_message).toEqual("No errors");
  expect(resp).toHaveProperty("test_mode");
  expect(resp).toHaveProperty("major");
  expect(resp).toHaveProperty("minor");
  expect(resp).toHaveProperty("patch");
  expect(resp.test_mode).toEqual(false);
});

test("publicKey", async () => {
  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  // Derivation path. First 3 items are automatically hardened!
  const path = [44, 931, 0, 0, 0];

  const resp = await app.publicKey(path);

  expect(resp.return_code).toEqual(ERROR_CODE.NoError);
  expect(resp.error_message).toEqual("No errors");
  expect(resp).toHaveProperty("compressed_pk");
  expect(resp.compressed_pk.length).toEqual(33);
  expect(resp.compressed_pk.toString("hex")).toEqual(
    "0264bff60f78aae3326b4fc3e16a7c48d0158bfe175fa4a9361190a42565662ae6",
  );
});

test("getAddressAndPubKey", async () => {
  jest.setTimeout(60000);

  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  // Derivation path. First 3 items are automatically hardened!
  const path = [44, 931, 0, 0, 0];
  const resp = await app.getAddressAndPubKey(path, "thor");

  console.log(resp);

  expect(resp.return_code).toEqual(ERROR_CODE.NoError);
  expect(resp.error_message).toEqual("No errors");

  expect(resp).toHaveProperty("bech32_address");
  expect(resp).toHaveProperty("compressed_pk");

  expect(resp.bech32_address).toEqual("thor1mwyrp6lj85swy5e5g4hjlaacm33g6rw3p4qmq4");
  expect(resp.compressed_pk.length).toEqual(33);
});

test("showAddressAndPubKey", async () => {
  jest.setTimeout(60000);

  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  // Derivation path. First 3 items are automatically hardened!
  const path = [44, 931, 0, 0, 0];
  const resp = await app.showAddressAndPubKey(path, "thor");

  console.log(resp);

  expect(resp.return_code).toEqual(ERROR_CODE.NoError);
  expect(resp.error_message).toEqual("No errors");

  expect(resp).toHaveProperty("bech32_address");
  expect(resp).toHaveProperty("compressed_pk");

  expect(resp.bech32_address).toEqual("thor1mwyrp6lj85swy5e5g4hjlaacm33g6rw3p4qmq4");
  expect(resp.compressed_pk.length).toEqual(33);
});

test("appInfo", async () => {
  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  const resp = await app.appInfo();

  console.log(resp);

  expect(resp.return_code).toEqual(ERROR_CODE.NoError);
  expect(resp.error_message).toEqual("No errors");

  expect(resp).toHaveProperty("appName");
  expect(resp).toHaveProperty("appVersion");
  expect(resp).toHaveProperty("flagLen");
  expect(resp).toHaveProperty("flagsValue");
  expect(resp).toHaveProperty("flag_recovery");
  expect(resp).toHaveProperty("flag_signed_mcu_code");
  expect(resp).toHaveProperty("flag_onboarded");
  expect(resp).toHaveProperty("flag_pin_validated");
});

test("deviceInfo", async () => {
  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  const resp = await app.deviceInfo();

  console.log(resp);

  expect(resp.return_code).toEqual(ERROR_CODE.NoError);
  expect(resp.error_message).toEqual("No errors");

  expect(resp).toHaveProperty("targetId");
  expect(resp).toHaveProperty("seVersion");
  expect(resp).toHaveProperty("flag");
  expect(resp).toHaveProperty("mcuVersion");
});

test("sign_and_verify_MsgSend", async () => {
  jest.setTimeout(60000);

  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  // Derivation path. First 3 items are automatically hardened!
  const path = [44, 931, 0, 0, 0];
  const message = String.raw`{"account_number":"588","chain_id":"thorchain","fee":{"amount":[],"gas":"2000000"},"memo":"TestMemo","msgs":[{"type":"thorchain/MsgSend","value":{"amount":[{"amount":"150000000","denom":"rune"}],"from_address":"tthor1c648xgpter9xffhmcqvs7lzd7hxh0prgv5t5gp","to_address":"tthor10xgrknu44d83qr4s4uw56cqxg0hsev5e68lc9z"}}],"sequence":"5"}`;

  const responsePk = await app.publicKey(path);
  console.log(responsePk);
  expect(responsePk.return_code).toEqual(ERROR_CODE.NoError);
  expect(responsePk.error_message).toEqual("No errors");

  const responseSign = await app.sign(path, message);
  console.log(responseSign);
  expect(responseSign.return_code).toEqual(ERROR_CODE.NoError);
  expect(responseSign.error_message).toEqual("No errors");

  // Check signature is valid
  const hash = crypto.createHash("sha256");
  const msgHash = hash.update(message).digest();

  const signatureDER = responseSign.signature;
  const signature = secp256k1.signatureImport(signatureDER);
  const signatureOk = secp256k1.verify(msgHash, signature, responsePk.compressed_pk);
  expect(signatureOk).toEqual(true);
});

test("sign_and_verify_MsgDeposit", async () => {
  jest.setTimeout(60000);

  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  // Derivation path. First 3 items are automatically hardened!
  const path = [44, 931, 0, 0, 0];
  const message = String.raw`{"account_number":"588","chain_id":"thorchain","fee":{"amount":[],"gas":"10000000"},"memo":"","msgs":[{"type":"thorchain/MsgDeposit","value":{"coins":[{"amount":"330000000","asset":"THOR.RUNE"}],"memo":"SWAP:BNB.BNB:tbnb1qk2m905ypazwfau9cn0qnr4c4yxz63v9u9md20:","signer":"tthor1c648xgpter9xffhmcqvs7lzd7hxh0prgv5t5gp"}}],"sequence":"6"}`;

  const responsePk = await app.publicKey(path);
  const responseSign = await app.sign(path, message);

  console.log(responsePk);
  console.log(responseSign);

  expect(responsePk.return_code).toEqual(ERROR_CODE.NoError);
  expect(responsePk.error_message).toEqual("No errors");
  expect(responseSign.return_code).toEqual(ERROR_CODE.NoError);
  expect(responseSign.error_message).toEqual("No errors");

  // Check signature is valid
  const hash = crypto.createHash("sha256");
  const msgHash = hash.update(message).digest();

  const signatureDER = responseSign.signature;
  const signature = secp256k1.signatureImport(signatureDER);
  const signatureOk = secp256k1.verify(msgHash, signature, responsePk.compressed_pk);
  expect(signatureOk).toEqual(true);
});

test("sign_invalid", async () => {
  jest.setTimeout(60000);

  const transport = await TransportNodeHid.create(1000);
  const app = new THORChainApp(transport);

  const path = [44, 931, 0, 0, 0]; // Derivation path. First 3 items are automatically hardened!
  const invalidMessage =
    '{"chain_id":"thorchain","fee":{"amount":[],"gas":"10000000"},"memo":"","msgs":[{"type":"thorchain/MsgDeposit","value":{"coins":[{"amount":"330000000","asset":"THOR.RUNE"}],"memo":"SWAP:BNB.BNB:tbnb1qk2m905ypazwfau9cn0qnr4c4yxz63v9u9md20:","signer":"tthor1c648xgpter9xffhmcqvs7lzd7hxh0prgv5t5gp"}}],"sequence":"6"}';

  const responseSign = await app.sign(path, invalidMessage);

  console.log(responseSign);

  switch (app.versionResponse.major) {
    case 2:
      expect(responseSign.return_code).toEqual(0x6984);
      expect(responseSign.error_message).toEqual("Data is invalid : JSON Missing account number");
      break;
    default:
      console.log("Version not supported");
      expect(false).toEqual(true);
  }
});


tests/e2e/basic.spec.js
import THORChainApp from "index.js";
import TransportNodeHid from "@ledgerhq/hw-transport-node-hid";
import { expect } from "jest";
import secp256k1 from "secp256k1/elliptic";
import crypto from "crypto";
import { ERROR_CODE } from "../../src/common";

require("dotenv").config({ path: "./.e2e.env" });

const { THOR_ADDRESS = "unknown address", THOR_PUB_KEY = "unknown pub key" } = process.env;

const JEST_TIMEOUT = 60000;
// Use same time out for all tests in this file
jest.setTimeout(JEST_TIMEOUT);

describe("THORChainApp e2e", () => {
  let transport;
  let app;
  beforeAll(async () => {
    transport = await TransportNodeHid.create(1000);
    app = new THORChainApp(transport);
  });

  afterAll(async () => {
    transport = null;
    app = null;
  });

  it("getVersion", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****
    const resp = await app.getVersion();

    expect(resp.return_code).toEqual(ERROR_CODE.NoError);
    expect(resp.error_message).toEqual("No errors");
    expect(resp).toHaveProperty("test_mode");
    expect(resp).toHaveProperty("major");
    expect(resp).toHaveProperty("minor");
    expect(resp).toHaveProperty("patch");
    expect(resp.test_mode).toEqual(false);
  });

  it("publicKey", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****

    // Derivation path. First 3 items are automatically hardened!
    const path = [44, 931, 0, 0, 0];

    const resp = await app.publicKey(path);

    // console.log(resp);

    expect(resp.return_code).toEqual(ERROR_CODE.NoError);
    expect(resp.error_message).toEqual("No errors");
    expect(resp).toHaveProperty("compressed_pk");
    expect(resp.compressed_pk.length).toEqual(33);
    expect(resp.compressed_pk.toString("hex")).toEqual(THOR_PUB_KEY);
  });

  it("getAddressAndPubKey", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****

    // Derivation path. First 3 items are automatically hardened!
    const path = [44, 931, 0, 0, 0];
    const resp = await app.getAddressAndPubKey(path, "thor");

    // console.log(resp);

    expect(resp.return_code).toEqual(ERROR_CODE.NoError);
    expect(resp.error_message).toEqual("No errors");

    expect(resp).toHaveProperty("bech32_address");
    expect(resp).toHaveProperty("compressed_pk");

    expect(resp.bech32_address).toEqual(THOR_ADDRESS);
    expect(resp.compressed_pk.length).toEqual(33);
  });

  it("appInfo", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****
    const resp = await app.appInfo();

    // console.log(resp);

    expect(resp.return_code).toEqual(ERROR_CODE.NoError);
    expect(resp.error_message).toEqual("No errors");

    expect(resp).toHaveProperty("appName");
    expect(resp).toHaveProperty("appVersion");
    expect(resp).toHaveProperty("flagLen");
    expect(resp).toHaveProperty("flagsValue");
    expect(resp).toHaveProperty("flag_recovery");
    expect(resp).toHaveProperty("flag_signed_mcu_code");
    expect(resp).toHaveProperty("flag_onboarded");
    expect(resp).toHaveProperty("flag_pin_validated");
  });

  it("deviceInfo", async () => {
    // *****
    // Requirements: Open `Dashboard` on Ledger (but not `THORChain` app)
    // *****
    const resp = await app.deviceInfo();

    // console.log(resp);

    expect(resp.return_code).toEqual(ERROR_CODE.NoError);
    expect(resp.error_message).toEqual("No errors");

    expect(resp).toHaveProperty("targetId");
    expect(resp).toHaveProperty("seVersion");
    expect(resp).toHaveProperty("flag");
    expect(resp).toHaveProperty("mcuVersion");
  });

  it("sign_and_verify_MsgSend", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger and follow instructions
    // *****

    // Derivation path. First 3 items are automatically hardened!
    const path = [44, 931, 0, 0, 0];
    const message = String.raw`{"account_number":"588","chain_id":"thorchain","fee":{"amount":[],"gas":"2000000"},"memo":"TestMemo","msgs":[{"type":"thorchain/MsgSend","value":{"amount":[{"amount":"150000000","denom":"rune"}],"from_address":"tthor1c648xgpter9xffhmcqvs7lzd7hxh0prgv5t5gp","to_address":"tthor10xgrknu44d83qr4s4uw56cqxg0hsev5e68lc9z"}}],"sequence":"5"}`;

    const responsePk = await app.publicKey(path);
    // console.log(responsePk);
    expect(responsePk.return_code).toEqual(ERROR_CODE.NoError);
    expect(responsePk.error_message).toEqual("No errors");

    const responseSign = await app.sign(path, message);
    // console.log(responseSign);
    expect(responseSign.return_code).toEqual(ERROR_CODE.NoError);
    expect(responseSign.error_message).toEqual("No errors");

    // Check signature is valid
    const hash = crypto.createHash("sha256");
    const msgHash = hash.update(message).digest();

    const signatureDER = responseSign.signature;
    const signature = secp256k1.signatureImport(signatureDER);
    // console.log(`signature`, signature);
    const signatureOk = secp256k1.ecdsaVerify(signature, msgHash, responsePk.compressed_pk);
    // console.log('signatureOks', signatureOk);
    expect(signatureOk).toEqual(true);
  });

  it("sign_invalid", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger and follow instructions
    // *****

    const path = [44, 931, 0, 0, 0]; // Derivation path. First 3 items are automatically hardened!
    const invalidMessage =
      '{"chain_id":"thorchain","fee":{"amount":[],"gas":"10000000"},"memo":"","msgs":[{"type":"thorchain/MsgDeposit","value":{"coins":[{"amount":"330000000","asset":"THOR.RUNE"}],"memo":"SWAP:BNB.BNB:tbnb1qk2m905ypazwfau9cn0qnr4c4yxz63v9u9md20:","signer":"tthor1c648xgpter9xffhmcqvs7lzd7hxh0prgv5t5gp"}}],"sequence":"6"}';

    const responseSign = await app.sign(path, invalidMessage);

    // console.log(responseSign);

    switch (app.versionResponse.major) {
      case 2:
        expect(responseSign.return_code).toEqual(0x6984);
        expect(responseSign.error_message).toEqual("Data is invalid : JSON Missing account number");
        break;
      default:
        console.log("Version not supported");
        expect(false).toEqual(true);
    }
  });
});


vue.config.js
const fs = require("fs");
const webpack = require('webpack');

module.exports = {

  configureWebpack: {
    resolve: {
      fallback: {
        crypto: require.resolve('crypto-browserify'),
        stream: require.resolve("stream-browserify")
        // Other fallbacks if necessary...
      },
    },
    plugins: [
      new webpack.ProvidePlugin({
        Buffer: ['buffer', 'Buffer'],
        // Other global variables if necessary...
      }),
    ],
    // Any additional Webpack configurations...
  },
  chainWebpack: config => {
    config.plugin('provide').use(webpack.ProvidePlugin, [{
      Buffer: ['buffer', 'Buffer'],
    }])
  },
  devServer: {
    https: {
      key: fs.readFileSync("certs/server.key"),
      cert: fs.readFileSync("certs/server.cert"),
      ca: fs.readFileSync("certs/cert.pem"),
    },
  },

};


src/helperV1.js
import { CLA, ERROR_CODE, errorCodeToString, INS, processErrorResponse } from "./common";

export function serializePathv1(path) {
  if (path == null || path.length < 3) {
    throw new Error("Invalid path.");
  }
  if (path.length > 10) {
    throw new Error("Invalid path. Length should be <= 10");
  }
  const buf = Buffer.alloc(1 + 4 * path.length);
  buf.writeUInt8(path.length, 0);
  for (let i = 0; i < path.length; i += 1) {
    let v = path[i];
    if (i < 3) {
      // eslint-disable-next-line no-bitwise
      v |= 0x80000000; // Harden
    }
    buf.writeInt32LE(v, 1 + i * 4);
  }
  return buf;
}

export async function signSendChunkv1(app, chunkIdx, chunkNum, chunk) {
  return app.transport
    .send(CLA, INS.SIGN_SECP256K1, chunkIdx, chunkNum, chunk, [ERROR_CODE.NoError, 0x6984, 0x6a80])
    .then((response) => {
      const errorCodeData = response.slice(-2);
      const returnCode = errorCodeData[0] * 256 + errorCodeData[1];
      let errorMessage = errorCodeToString(returnCode);

      if (returnCode === 0x6a80 || returnCode === 0x6984) {
        errorMessage = `${errorMessage} : ${response.slice(0, response.length - 2).toString("ascii")}`;
      }

      let signature = null;
      if (response.length > 2) {
        signature = response.slice(0, response.length - 2);
      }

      return {
        signature,
        return_code: returnCode,
        error_message: errorMessage,
      };
    }, processErrorResponse);
}

function compressPublicKey(publicKey) {
  if (publicKey.length !== 65) {
    throw new Error("decompressed public key length should be 65 bytes");
  }
  const y = publicKey.slice(33, 65);
  // eslint-disable-next-line no-bitwise
  const z = Buffer.from([2 + (y[y.length - 1] & 1)]);
  return Buffer.concat([z, publicKey.slice(1, 33)]);
}

export async function publicKeyv1(app, data) {
  return app.transport
    .send(CLA, INS.INS_PUBLIC_KEY_SECP256K1, 0, 0, data, [ERROR_CODE.NoError])
    .then((response) => {
      const errorCodeData = response.slice(-2);
      const returnCode = errorCodeData[0] * 256 + errorCodeData[1];
      const pk = Buffer.from(response.slice(0, 65));

      return {
        pk,
        compressed_pk: compressPublicKey(pk),
        return_code: returnCode,
        error_message: errorCodeToString(returnCode),
      };
    }, processErrorResponse);
}


src/signatureExtract.ts
/**
* Takes raw output from Ledger device which is TLV encoded in the format "0x30 L 0x02 Lr r 0x02 Ls s"
* where L is length of entire message (after 2nd byte), Lr is length of r and Ls is length of s.
* This function extracts r and s, ensures they are 32 bytes each, joins into a single 64 byte Array.
* This value returned should be base64 encoded to a string to add to the JSON 'signatures' to send to RPC.
* 
* Example raw sig:
* [ 48, 69, 2, 33, 
*   0, 161, 207, 217, 127, 151, 190, 23, 98, 217, 186, 108, 82, 102, 19, 222, 20, 171, 6, 240, 134, 195, 251, 98, 190, 246, 228, 243, 215, 95, 166, 121, 165, 
*   2, 32, 
*   69, 188, 53,  213, 24, 9, 191, 90, 244, 21, 213, 146, 240, 109, 156, 221, 247, 63, 131, 52, 150, 253, 199, 153, 132, 76, 91, 239, 28, 254, 68, 80 ]
* 
*   48 69  -- length is 69
*   2 33   -- length of r is 33 bytes 
*   <33 bytes of r>  -- (the leading zero here can be dropped)
*   2 32   -- length of s is 32 bytes
*   <32 bytes of s>
*/
export function extractSignatureFromTLV(signatureArray : number[]): number[] {

    // Check Type Length Value encoding
    if (signatureArray.length < 64) { throw Error('Invalid Signature: Too short') }
    if (signatureArray[0] != 0x30) { throw Error('Invalid Ledger Signature TLV encoding: expected first byte 0x30') }
    if (signatureArray[1] + 2 != signatureArray.length) { throw Error('Invalid Signature: signature length does not match TLV') }
    if (signatureArray[2] != 0x02) { throw Error('Invalid Ledger Signature TLV encoding: expected length type 0x02') }

    // r signature
    const rLength = signatureArray[3]
    var rSignature = signatureArray.slice(4, rLength + 4)
    // Drop leading zero on some 'r' signatures that are 33 bytes.
    if (rSignature.length == 33 && rSignature[0] == 0) {
        rSignature = rSignature.slice(1,33)
    } else if (rSignature.length == 33) {
        throw Error('Invalid signature: "r" too long')
    }
    // add leading zero's to pad to 32 bytes
    while (rSignature.length < 32) { rSignature.unshift(0) }

    // s signature
    if (signatureArray[rLength + 4] != 0x02) { throw Error('Invalid Ledger Signature TLV encoding: expected length type 0x02') }
    const sLength = signatureArray[rLength + 5]
    if (4 + rLength + 2 + sLength != signatureArray.length) { throw Error('Invalid Ledger Signature: TLV byte lengths do not match message length') }
    var sSignature = signatureArray.slice(rLength + 6, signatureArray.length)
    // Drop leading zero on 's' signatures that are 33 bytes. This shouldn't occur since ledger signs using "Small s" math. But just to be sure...
    if (sSignature.length == 33 && sSignature[0] == 0) {
        sSignature = sSignature.slice(1,33)
    } else if (sSignature.length == 33) {
        throw Error('Invalid signature: "s" too long')
    }
    // add leading zero's to pad to 32 bytes
    while (sSignature.length < 32) { sSignature.unshift(0) }

    if (rSignature.length != 32 || sSignature.length != 32) { throw Error('Invalid signatures: must be 32 bytes each') }

    return rSignature.concat(sSignature)
}


src/index.js
/** ******************************************************************************
 *  (c) 2019 ZondaX GmbH
 *  (c) 2016-2017 Ledger
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 ******************************************************************************* */

import crypto from "crypto";
import Ripemd160 from "ripemd160";
import { bech32 } from "bech32";
import { publicKeyv2, serializePathv2, signSendChunkv2 } from "./helperV2";
import {
  APP_KEY,
  CHUNK_SIZE,
  CLA,
  INS,
  errorCodeToString,
  getVersion,
  processErrorResponse,
  ERROR_CODE,
  P1_VALUES,
} from "./common";

export default class THORChainApp {
  constructor(transport, scrambleKey = APP_KEY) {
    if (!transport) {
      throw new Error("Transport has not been defined");
    }

    this.transport = transport;
    transport.decorateAppAPIMethods(
      this,
      ["getVersion", "sign", "getAddressAndPubKey", "appInfo", "deviceInfo", "getBech32FromPK"],
      scrambleKey,
    );
  }

  static serializeHRP(hrp) {
    if (hrp == null || hrp.length < 3 || hrp.length > 83) {
      throw new Error("Invalid HRP");
    }
    const buf = Buffer.alloc(1 + hrp.length);
    buf.writeUInt8(hrp.length, 0);
    buf.write(hrp, 1);
    return buf;
  }

  static getBech32FromPK(hrp, pk) {
    if (pk.length !== 33) {
      throw new Error("expected compressed public key [31 bytes]");
    }

    const hashSha256 = crypto.createHash("sha256").update(pk).digest();
    const hashRip = new Ripemd160().update(hashSha256).digest();
    const words = bech32.toWords(hashRip)
    return bech32.encode(hrp, words);
  }

  async serializePath(path) {
    this.versionResponse = await getVersion(this.transport);

    if (this.versionResponse.return_code !== ERROR_CODE.NoError) {
      throw this.versionResponse;
    }

    switch (this.versionResponse.major) {
      case 2:
        return serializePathv2(path);
      default:
        return {
          return_code: 0x6400,
          error_message: "App Version is not supported",
        };
    }
  }

  async signGetChunks(path, message) {
    const serializedPath = await this.serializePath(path);

    const chunks = [];
    chunks.push(serializedPath);
    const buffer = Buffer.from(message);

    for (let i = 0; i < buffer.length; i += CHUNK_SIZE) {
      let end = i + CHUNK_SIZE;
      if (i > buffer.length) {
        end = buffer.length;
      }
      chunks.push(buffer.slice(i, end));
    }

    return chunks;
  }

  async getVersion() {
    try {
      this.versionResponse = await getVersion(this.transport);
      return this.versionResponse;
    } catch (e) {
      return processErrorResponse(e);
    }
  }

  async appInfo() {
    return this.transport.send(0xb0, 0x01, 0, 0).then((response) => {
      const errorCodeData = response.slice(-2);
      const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

      const result = {};

      let appName = "err";
      let appVersion = "err";
      let flagLen = 0;
      let flagsValue = 0;

      if (response[0] !== 1) {
        // Ledger responds with format ID 1. There is no spec for any format != 1
        result.error_message = "response format ID not recognized";
        result.return_code = 0x9001;
      } else {
        const appNameLen = response[1];
        appName = response.slice(2, 2 + appNameLen).toString("ascii");
        let idx = 2 + appNameLen;
        const appVersionLen = response[idx];
        idx += 1;
        appVersion = response.slice(idx, idx + appVersionLen).toString("ascii");
        idx += appVersionLen;
        const appFlagsLen = response[idx];
        idx += 1;
        flagLen = appFlagsLen;
        flagsValue = response[idx];
      }

      return {
        return_code: returnCode,
        error_message: errorCodeToString(returnCode),
        // //
        appName,
        appVersion,
        flagLen,
        flagsValue,
        // eslint-disable-next-line no-bitwise
        flag_recovery: (flagsValue & 1) !== 0,
        // eslint-disable-next-line no-bitwise
        flag_signed_mcu_code: (flagsValue & 2) !== 0,
        // eslint-disable-next-line no-bitwise
        flag_onboarded: (flagsValue & 4) !== 0,
        // eslint-disable-next-line no-bitwise
        flag_pin_validated: (flagsValue & 128) !== 0,
      };
    }, processErrorResponse);
  }

  async deviceInfo() {
    return this.transport
      .send(0xe0, 0x01, 0, 0, Buffer.from([]), [ERROR_CODE.NoError, 0x6e00])
      .then((response) => {
        const errorCodeData = response.slice(-2);
        const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

        if (returnCode === 0x6e00) {
          return {
            return_code: returnCode,
            error_message: "This command is only available in the Dashboard",
          };
        }

        const targetId = response.slice(0, 4).toString("hex");

        let pos = 4;
        const secureElementVersionLen = response[pos];
        pos += 1;
        const seVersion = response.slice(pos, pos + secureElementVersionLen).toString();
        pos += secureElementVersionLen;

        const flagsLen = response[pos];
        pos += 1;
        const flag = response.slice(pos, pos + flagsLen).toString("hex");
        pos += flagsLen;

        const mcuVersionLen = response[pos];
        pos += 1;
        // Patch issue in mcu version
        let tmp = response.slice(pos, pos + mcuVersionLen);
        if (tmp[mcuVersionLen - 1] === 0) {
          tmp = response.slice(pos, pos + mcuVersionLen - 1);
        }
        const mcuVersion = tmp.toString();

        return {
          return_code: returnCode,
          error_message: errorCodeToString(returnCode),
          // //
          targetId,
          seVersion,
          flag,
          mcuVersion,
        };
      }, processErrorResponse);
  }

  async publicKey(path) {
    try {
      const serializedPath = await this.serializePath(path);

      switch (this.versionResponse.major) {
        case 2: {
          const data = Buffer.concat([THORChainApp.serializeHRP("thor"), serializedPath]);
          return publicKeyv2(this, data);
        }
        default:
          return {
            return_code: 0x6400,
            error_message: "App Version is not supported",
          };
      }
    } catch (e) {
      return processErrorResponse(e);
    }
  }

  async getAddressAndPubKey(path, hrp) {
    try {
      return this.serializePath(path)
        .then((serializedPath) => {
          const data = Buffer.concat([THORChainApp.serializeHRP(hrp), serializedPath]);
          return this.transport
            .send(CLA, INS.GET_ADDR_SECP256K1, P1_VALUES.ONLY_RETRIEVE, 0, data, [ERROR_CODE.NoError])
            .then((response) => {
              const errorCodeData = response.slice(-2);
              const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

              const compressedPk = Buffer.from(response.slice(0, 33));
              const bech32Address = Buffer.from(response.slice(33, -2)).toString();

              return {
                bech32_address: bech32Address,
                compressed_pk: compressedPk,
                return_code: returnCode,
                error_message: errorCodeToString(returnCode),
              };
            }, processErrorResponse);
        })
        .catch((err) => processErrorResponse(err));
    } catch (e) {
      return processErrorResponse(e);
    }
  }

  async showAddressAndPubKey(path, hrp) {
    try {
      return this.serializePath(path)
        .then((serializedPath) => {
          const data = Buffer.concat([THORChainApp.serializeHRP(hrp), serializedPath]);
          return this.transport
            .send(CLA, INS.GET_ADDR_SECP256K1, P1_VALUES.SHOW_ADDRESS_IN_DEVICE, 0, data, [
              ERROR_CODE.NoError,
            ])
            .then((response) => {
              const errorCodeData = response.slice(-2);
              const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

              const compressedPk = Buffer.from(response.slice(0, 33));
              const bech32Address = Buffer.from(response.slice(33, -2)).toString();

              return {
                bech32_address: bech32Address,
                compressed_pk: compressedPk,
                return_code: returnCode,
                error_message: errorCodeToString(returnCode),
              };
            }, processErrorResponse);
        })
        .catch((err) => processErrorResponse(err));
    } catch (e) {
      return processErrorResponse(e);
    }
  }

  async signSendChunk(chunkIdx, chunkNum, chunk) {
    switch (this.versionResponse.major) {
      case 2:
        return signSendChunkv2(this, chunkIdx, chunkNum, chunk);
      default:
        return {
          return_code: 0x6400,
          error_message: "App Version is not supported",
        };
    }
  }

  async sign(path, message) {
    return this.signGetChunks(path, message).then((chunks) => {
      return this.signSendChunk(1, chunks.length, chunks[0], [ERROR_CODE.NoError]).then(async (response) => {
        let result = {
          return_code: response.return_code,
          error_message: response.error_message,
          signature: null,
        };

        for (let i = 1; i < chunks.length; i += 1) {
          // eslint-disable-next-line no-await-in-loop
          result = await this.signSendChunk(1 + i, chunks.length, chunks[i]);
          if (result.return_code !== ERROR_CODE.NoError) {
            break;
          }
        }

        return {
          return_code: result.return_code,
          error_message: result.error_message,
          // ///
          signature: result.signature,
        };
      }, processErrorResponse);
    }, processErrorResponse);
  }
}
